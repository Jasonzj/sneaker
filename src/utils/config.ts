const interfaceName = {
  stockx: 'stockx',
  goat: 'goat',
  dewu: 'dewu',
  db: 'db',
}

const defaultInterface = interfaceName.dewu

export default {
  title: 'Sneakers App',
  bannerText: 'Search for Shoes and Integration Price For All Size',
  openPages: ['/sign_in', '/sign_up'],
  homePage: '/trending/dewu',
  apiPrefix: '/api/v1',
  interfaceName,
  defaultInterface,
}
