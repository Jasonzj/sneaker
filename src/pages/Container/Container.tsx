import React from 'react'
import { Switch, Route, Redirect, useRouteMatch } from 'react-router-dom'
import { CSSTransition, SwitchTransition } from 'react-transition-group'

// components
import Product from '../Product'
import NoMatch from '../NoMatch/'
import Header from '../../components/Header'
import Banner from '../../components/Banner'
import SignIn from '../SignIn'
import SignUp from '../SignUp'

// config
import config from '../../utils/config'
const { homePage, openPages } = config

const openPagesMap = {
  '/sign_in': <Route exact path='/sign_in' component={SignIn} />,
  '/sign_up': <Route exact path='/sign_up' component={SignUp} />,
}

const Container: React.FC = () => {
  const match = useRouteMatch({
    path: ['/sign_in', '/sign_up'],
    exact: true,
  })

  if (match && openPages && openPages.includes(match.path)) {
    return (
      <SwitchTransition mode='out-in'>
        <CSSTransition key={match.path} timeout={1000} in classNames='sign-ani'>
          {openPagesMap[match.path as '/sign_in' | '/sign_up']}
        </CSSTransition>
      </SwitchTransition>
    )
  }

  return (
    <div className='App'>
      <Header />
      <Banner />
      <Switch>
        <Route exact path='/' render={() => <Redirect to={homePage} />} />
        <Route exact path='/trending/:siteName' component={Product} />
        <Route exact path='/search/:siteName/:key' component={Product} />
        <Route exact path='/following' component={Product} />
        <Route component={NoMatch} />
      </Switch>
    </div>
  )
}

export default Container
