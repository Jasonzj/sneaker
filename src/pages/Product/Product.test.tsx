import React from 'react'
import user from '@testing-library/user-event'
import { render, screen } from '../../tests/routerRender'
import Container from '../Container/Container'
import { rest, server } from '../../tests/testSever'
import { product_error } from '../../tests/test.mock'
import config from '../../utils/config'
const { interfaceName } = config

const setup = () => {
  const { unmount, asFragment } = render(<Container />)

  const Search_Title = screen.getByTitle('search_title')
  const Search_Input = screen.getByTitle('search_input')
  const Search_Button = screen.getByTitle('search_button')
  const Search_Select = screen.getByTitle('search_interface_select')

  return { unmount, asFragment, Search_Title, Search_Input, Search_Button, Search_Select }
}

test('should display Trending(dewu) in first', async () => {
  const { asFragment, Search_Title } = setup()

  await screen.findByTestId('product_card')

  expect(Search_Title).toHaveTextContent('Trending(dewu)')
  expect(asFragment()).toMatchSnapshot()
})

test('keyword empty not allow search', async () => {
  const { Search_Title, Search_Button } = setup()

  await screen.findByTestId('product_card')
  user.click(Search_Button)

  expect(Search_Title).toHaveTextContent('Trending(dewu)')
})

test('title should display keyword after search', async () => {
  const { asFragment, Search_Title, Search_Input, Search_Button, Search_Select } = setup()

  user.type(Search_Input, 'Air')
  user.selectOptions(Search_Select, interfaceName.goat)
  user.click(Search_Button)

  await screen.findByTestId('product_card')

  expect(Search_Title).toHaveTextContent('Search goat for (Air)')
  expect(asFragment()).toMatchSnapshot()
})

test('select onChange should search again', async () => {
  const { asFragment, Search_Title, Search_Button, Search_Select } = setup()

  user.selectOptions(Search_Select, interfaceName.stockx)
  user.click(Search_Button)

  await screen.findByTestId('product_card')

  expect(Search_Title).toHaveTextContent('Search stockx for (Air)')
  expect(asFragment()).toMatchSnapshot()
})

test('keyword onChange should search again', async () => {
  const { asFragment, Search_Title, Search_Button, Search_Input } = setup()

  user.clear(Search_Input)
  user.type(Search_Input, 'Nike')
  user.click(Search_Button)

  await screen.findByTestId('product_card')

  expect(Search_Title).toHaveTextContent('Search stockx for (Nike)')
  expect(asFragment()).toMatchSnapshot()
})

test('trending button onClick should jump correct', async () => {
  const { asFragment, Search_Title } = setup()

  user.click(screen.getByText('Goat Trending'))

  await screen.findByTestId('product_card')

  expect(screen.getByTitle('search_title')).toHaveTextContent('Trending(goat)')
  expect(asFragment()).toMatchSnapshot()

  user.click(screen.getByText('Dewu Trending'))

  await screen.findByTestId('product_card')

  expect(Search_Title).toHaveTextContent('Trending(dewu)')
  expect(asFragment()).toMatchSnapshot()
})

test('should handle server error', async () => {
  server.use(
    rest.get('/api/v1/trending/dewu', (req, res, ctx) => {
      return res(ctx.status(500))
    }),
  )

  const { asFragment } = setup()

  expect(await screen.findByText('Search Fail!')).toBeInTheDocument()
  expect(asFragment()).toMatchSnapshot()

  server.use(
    rest.get('/api/v1/trending/dewu', (req, res, ctx) => {
      return res(ctx.status(200), ctx.json(product_error))
    }),
  )

  user.click(screen.getByText('Goat Trending'))
  user.click(screen.getByText('Dewu Trending'))

  expect(await screen.findByText('Search Fail!')).toBeInTheDocument()
  expect(asFragment()).toMatchSnapshot()
})

test('should show correctly productDetail Model', async () => {
  setup()

  user.click(await screen.findByTestId('product_card'))

  expect(await screen.findByText('US Size')).toBeInTheDocument()
})

test('should show correctly following button', async () => {
  const { unmount, asFragment } = setup()

  user.click(await screen.findByTestId('product_card'))

  expect(await screen.findByText('unfollow')).toBeInTheDocument()
  expect(asFragment()).toMatchSnapshot()

  server.use(
    rest.get('/api/v1/user/following', (req, res, ctx) => {
      return res(ctx.status(500))
    }),
  )

  unmount()
  const { asFragment: asFragment2 } = setup()

  user.click(await screen.findByTestId('product_card'))

  expect(await screen.findByText('follow')).toBeInTheDocument()
  expect(asFragment2()).toMatchSnapshot()

  // prevent tests leaking
  await screen.findByText('Show Size Chart')
})

// test('should handle server error in productDeatil', async () => {
//   server.use(
//     rest.get('/api/v1/prices/554724-058', (req, res, ctx) => {
//       return res(ctx.status(500))
//     }),
//   )

//   setup()

//   const Product_Card = await screen.findByTestId('product_card')

//   user.click(Product_Card)

//   expect(await screen.findByText('No BaseProperties')).toBeInTheDocument()

//   server.use(
//     rest.get('/api/v1/prices/554724-058', (req, res, ctx) => {
//       return res(ctx.json(prices_error))
//     }),
//   )

//   user.click(Product_Card)

//   expect(await screen.findByText('No BaseProperties')).toBeInTheDocument()
// })
