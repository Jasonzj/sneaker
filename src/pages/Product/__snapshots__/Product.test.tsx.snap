// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header
      class="header"
    >
      <div
        class="logoBox"
      >
        <h1>
          Sneakers App
        </h1>
        <img
          alt="logo"
          class="logo"
          src="logo.png"
        />
      </div>
      <nav
        class="nav"
      >
        <a
          href="#/trending/dewu"
        >
          Dewu Trending
        </a>
        <a
          href="#/trending/goat"
        >
          Goat Trending
        </a>
        <div
          class="main"
        >
          <div
            class="content"
          >
            <div
              class="avatar"
            >
              <img
                src="https://api.prodless.com/avatar.png"
              />
            </div>
          </div>
        </div>
      </nav>
    </header>
    <div
      class="banner"
    >
      <div
        class="introduce"
      >
        Search for Shoes and Integration Price For All Size
      </div>
      <div
        class="searchBox"
      >
        <input
          class="searchInput"
          title="search_input"
          type="text"
          value=""
        />
        <button
          class="searchButton"
          title="search_button"
        />
        <div
          class="selectBox"
        >
          <span>
            Search Interface: 
          </span>
          <select
            class="searchInterface"
            title="search_interface_select"
          >
            <option
              value="dewu"
            >
              Dewu
            </option>
            <option
              value="stockx"
            >
              StockX
            </option>
            <option
              value="goat"
            >
              Goat
            </option>
            <option
              value="db"
            >
              DB
            </option>
          </select>
          <span>
            Search Suggestions Interface: 
          </span>
          <select
            class="searchInterface"
          >
            <option
              value="dewu"
            >
              Dewu
            </option>
            <option
              value="stockx"
            >
              StockX
            </option>
          </select>
        </div>
        <div
          class="searchSuggestion"
        />
      </div>
      <div
        class="iconBar"
      >
        <img
          alt="goat"
          class="logo"
          src="goat.png"
        />
        <img
          alt="dewu"
          class="logo"
          src="dewu.png"
        />
        <img
          alt="stockx"
          class="logo"
          src="stockx.png"
        />
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="header"
      >
        <h1
          class="title"
          title="search_title"
        >
          Trending(dewu)
        </h1>
      </div>
      <div
        class="box"
      >
        <div
          class="loader hidden"
          title="loader"
        >
          <div
            class="warpper"
          >
            <div
              class="inner"
            />
            <div
              class="text"
            >
              LOADING
            </div>
          </div>
        </div>
        <div
          class="card"
          data-testid="product_card"
        >
          <img
            alt="Air Jordan 1 Mid Shattered Backboard 扣碎篮板 篮球鞋"
            src="https://cdn.poizon.com/temp/Fn20HZPjmCnQk_VCNJ9C2HRg0xP5"
          />
          <div
            class="shoeName"
          >
            Air Jordan 1 Mid Shattered Backboard 扣碎篮板 篮球鞋
          </div>
          <div
            class="price"
          >
            $109
          </div>
          <div
            class="brand"
          >
            stockx
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
